cmd/notary/tuf.go

```diff
@@ -270,58 +272,92 @@ func verify(cmd *cobra.Command, args []string) {
-		fmt.Println("Please use a password manager to generate and store a good random passphrase.")
-		return "", errors.New("Passphrase too short")
-	}
-	return passphrase, nil
-}
+func getNotaryPassphraseRetriever() trustmanager.PassphraseRetriever {
+	userEnteredTargetsSnapshotsPass := false
+	targetsSnapshotsPass := ""
+	userEnteredRootsPass := false
+	rootsPass := ""
+
+	return func(keyID string, alias string, createNew bool, numAttempts int) (string, bool, error) {
+
+		// First, check if we have a password cached for this alias.
+		if numAttempts == 0 {
+			if userEnteredTargetsSnapshotsPass && (alias == "snapshot" || alias == "targets") {
+				return targetsSnapshotsPass, false, nil
+			}
+			if userEnteredRootsPass && (alias == "root") {
+				return rootsPass, false, nil
+			}
+		}
 
-func getPassphrase(confirm bool) ([]byte, error) {
-	if pass := os.Getenv("NOTARY_ROOT_PASSPHRASE"); pass != "" {
-		return []byte(pass), nil
-	}
+		if numAttempts > 3 && !createNew {
+			return "", true, errors.New("Too many attempts")
+		}
 
-	state, err := term.SaveState(0)
-	if err != nil {
-		return nil, err
-	}
-	term.DisableEcho(0, state)
-	defer term.RestoreTerminal(0, state)
+		state, err := term.SaveState(0)
+		if err != nil {
+			return "", false, err
+		}
+		term.DisableEcho(0, state)
+		defer term.RestoreTerminal(0, state)
 
-	stdin := bufio.NewReader(os.Stdin)
+		stdin := bufio.NewReader(os.Stdin)
 
-	fmt.Printf("Enter root key passphrase: ")
-	passphrase, err := stdin.ReadBytes('\n')
-	fmt.Println()
-	if err != nil {
-		return nil, err
-	}
-	passphrase = passphrase[0 : len(passphrase)-1]
+		if createNew {
+			fmt.Printf("Enter passphrase for new %s key with id %s: ", alias, keyID)
+		} else {
+			fmt.Printf("Enter key passphrase for %s key with id %s: ", alias, keyID)
+		}
 
-	if !confirm {
-		return passphrase, nil
-	}
+		passphrase, err := stdin.ReadBytes('\n')
+		fmt.Println()
+		if err != nil {
+			return "", false, err
+		}
+		passphrase = passphrase[0 : len(passphrase)-1]
+
+		if !createNew {
+			retPass := string(passphrase)
+			if alias == "snapshot" || alias == "targets" {
+				userEnteredTargetsSnapshotsPass = true
+				targetsSnapshotsPass = retPass
+			}
+			if alias == "root" {
+				userEnteredRootsPass = true
+				rootsPass = retPass
+			}
+			return string(passphrase), false, nil
+		}
 
-	fmt.Printf("Repeat root key passphrase: ")
-	confirmation, err := stdin.ReadBytes('\n')
-	fmt.Println()
-	if err != nil {
-		return nil, err
-	}
-	confirmation = confirmation[0 : len(confirmation)-1]
+		if len(passphrase) < 8 {
+			fmt.Println("Please use a password manager to generate and store a good random passphrase.")
+			return "", false, errors.New("Passphrase too short")
+		}
+
+		fmt.Printf("Repeat passphrase for new %s key with id %s: ", alias, keyID)
+		confirmation, err := stdin.ReadBytes('\n')
+		fmt.Println()
+		if err != nil {
+			return "", false, err
+		}
+		confirmation = confirmation[0 : len(confirmation)-1]
```

Comment:
Is len(confirmation) guaranteed to be positive here? I checked the docs for ReadBytes and I think the answer is yes, but I'd still err on the side of caution.